src/util/skiplist.c

```diff
@@ -0,0 +1,451 @@
+    /* store rank that is crossed to reach the insert position */
+    rank[i] = (i == (sl->level - 1)) ? 0 : rank[i + 1];
+    while (x->level[i].forward &&
+           sl->compare(x->level[i].forward->obj, obj, sl->cmpCtx) < 0) {
+      rank[i] += x->level[i].span;
+      x = x->level[i].forward;
+    }
+    update[i] = x;
+  }
+
+  /* If the element is already inside, append the value to the element. */
+  if (x->level[0].forward &&
+      sl->compare(x->level[0].forward->obj, obj, sl->cmpCtx) == 0) {
+    return skiplistNodeAppendValue(x->level[0].forward, val, sl->valcmp);
+  }
+
+  /* Add a new node with a random number of levels. */
+  level = skiplistRandomLevel();
+  if (level > sl->level) {
+    for (i = sl->level; i < level; i++) {
+      rank[i] = 0;
+      update[i] = sl->header;
+      update[i]->level[i].span = sl->length;
+    }
+    sl->level = level;
+  }
+  x = skiplistCreateNode(level, obj, val);
+  for (i = 0; i < level; i++) {
+    x->level[i].forward = update[i]->level[i].forward;
+    update[i]->level[i].forward = x;
+
+    /* update span covered by update[i] as x is inserted here */
+    x->level[i].span = update[i]->level[i].span - (rank[0] - rank[i]);
+    update[i]->level[i].span = (rank[0] - rank[i]) + 1;
+  }
+
+  /* increment span for untouched levels */
+  for (i = level; i < sl->level; i++) {
+    update[i]->level[i].span++;
+  }
+
+  x->backward = (update[0] == sl->header) ? NULL : update[0];
+  if (x->level[0].forward) {
+    x->level[0].forward->backward = x;
+  } else {
+    sl->tail = x;
+  }
+  sl->length++;
+
+  return x;
+}
+
+/*
+ * Internal function used by skiplistDelete, it needs an array of other
+ * skiplist nodes that point to the node to delete in order to update
+ * all the references of the node we are going to remove.
+ */
+void skiplistDeleteNode(skiplist *sl, skiplistNode *x, skiplistNode **update) {
+  // Free the node key if we specified a routine for it
+  if (sl->freeObj) sl->freeObj(x->obj);
+
+  int i;
+  for (i = 0; i < sl->level; i++) {
+    if (update[i]->level[i].forward == x) {
+      update[i]->level[i].span += x->level[i].span - 1;
+      update[i]->level[i].forward = x->level[i].forward;
+    } else {
+      update[i]->level[i].span -= 1;
+    }
+  }
+  if (x->level[0].forward) {
+    x->level[0].forward->backward = x->backward;
+  } else {
+    sl->tail = x->backward;
+  }
+  while (sl->level > 1 && sl->header->level[sl->level - 1].forward == NULL) {
+    sl->level--;
+  }
+  sl->length--;
+}
+
+/*
+ * Delete an element from the skiplist. If the `val` argument is provided,
+ * the single matching value will be found and deleted; otherwise, the skiplistNode
+ * representing the key matching `obj` and all of its elements will be deleted.
+ * If the operation is successful, 1 is returned, otherwise the target was not found and 0 is returned.
+ * The memory used to to represent skiplist keys and values is not freed.
+ */
+int skiplistDelete(skiplist *sl, void *obj, void *val) {
```

Comment:
Do we want to realloc vals where the ratio between numVals and valsAllocated is
numVals == 0.25*valsAllocated ?
as there are plenty of unused space